File: Programmers.co.kr/Level 3/자물쇠와 열쇠/solution.py
def solution(key, lock):
    M = len(key[0])         # key  is M x M array
    N = len(lock[0])        # lock is N x N array

    for rot_time in range(4):       # rotate key 3 times

        ## Shifting key
        for x in range(1, N + M + 1):         # shift in x axis
            for y in range(1, N + M + 1):     # shift in y axis

                ## XOR operation to elements of rotated & shifted key and lock
                result = True
                for i in range(N):
                    for j in range(N):
                        ## Case 1: rotated & shifted key and lock overlap
                        if (i in range( max(0, x - M), min(x, N) )) & (j in range( max(0, y - M), min(y, N) )):
                            result &= key[M - x + i][M - y + j] ^ lock[i][j]

                        ## Case 2: key and lock does not overlap
                        else:
                            result &= lock[i][j]
                
                ## Return True if result == True
                if result:
                    return True
        
        ## Rotate  key
        key = [[key[M - 1 - j][i] for j in range(M)] for i in range(M)]

    return False

File: Programmers.co.kr/Level 3/자물쇠와 열쇠/solution.cpp
#include <string>
#include <vector>

using namespace std;

bool solution(vector<vector<int>> key, vector<vector<int>> lock) {
    int M = key[0].size();        // key is  M x M array
    int N = lock[0].size();       // lock is N x N array
    
    // Rotate key 4 times.
    for(int rot_time = 0; rot_time < 4; rot_time++)
    {
        // Shift key.
        for(int x = 0; x < N + M + 1; x++)      // shifting in x axis
        {
            for(int y = 0; y < N + M + 1; y++)  // shifting in y axis
            {
                // XOR operation to elements of rotated & shifted key and lock.
                bool result = true;
                for(int i = 0; i < N; i++)
                {
                    for(int j = 0; j < N; j++)
                    {
                        // Case 1: rotated & shifted key and lock overlap
                        if( (i >= max(0, x - M)) & (i < min(x, N)) & (j >= max(0, y - M)) & (j < min(y, N)) )
                        {
                            result &= key[M - x + i][M - y + j] ^ lock[i][j];
                        }
                        
                        // Case 2: key and lock does not overlap
                        else result &= lock[i][j];
                    }
                }
                
                if(result) return true;
            }
                
        }
        
        // Rotate key.
        vector<vector<int>> rotated_key = key;
        for(int i = 0; i < M; i++)
        {
            for(int j = 0; j < M; j++)
            {
                rotated_key[i][j] = key[M - 1 - j][i];
            }
        }
                                     
        key = rotated_key;
    }
    
    return false;
}
